### PR TITLE
CONSOLE-3169: Publish console dynamic plugin extension and api docs

### PR DESCRIPTION
**Pull request by TheRealJon**
Manually publish dynamic plugin SDK API docs and add CI checks to ensure generated files stay up to date.

**`README.md`**
Update root README to provide easier navigation to package docs.

**`frontend/packages/console-dynamic-plugin-sdk/README.md`**
Update console dynamic plugin SDK README to provide easier navigation to API and Console Extension docs.

**`frontend/packages/console-dynamic-plugin-sdk/docs`**
Generate dynamic plugin docs in this dir so that they are not ignored by git and `/generated/` is not included in the GitHub Pages URL path.

**`frontend/packages/console-dynamic-plugin-sdk/scripts/generate-doc.ts`**
Refactor and simplify GitHub link generation logic

**`frontend/packages/console-dynamic-plugin-sdk/scripts/package-definitions.ts`**
Update location of generated docs so that they are copied into dist

**`test-frontend.sh`**
Reorder the checks in this file so that the least time-consuming and compute-intensive steps come first, meaning we will fail faster.

https://issues.redhat.com/browse/CONSOLE-3169

### CODE REVIEW

**Review by spadgett — COMMENTED** [review]
@TheRealJon If we check in the generated doc files, we should add something to CI that makes sure they're updated when things changed (similar to what we do with yarn.lock).

@vojtechszocs @christianvogt @bipuladh FYI

**Comment by TheRealJon** [comment]
@spadgett 

> @TheRealJon If we check in the generated doc files, we should add something to CI that makes sure they're updated when things changed (similar to what we do with yarn.lock).

Yeah, definitely. I was going to look into this next


> Can we remove the `generated` folder from gitignore in the SDK package?

I opted to move the docs out of the generated folder since those are the only files we need to check in. The other artifacts in the `generated` folder can still be ignored. I also thought it would be good to keep `/generated/` out of the URL path for GitHub Pages.

**Review by spadgett — APPROVED** [review]
/lgtm

**Comment by TheRealJon** [comment]
/hold

There seems to be an issue with the generation script. It's generating different source code links in CI and local environments.

**Comment by TheRealJon** [comment]
/hold cancel

I fixed the issue with the link generation logic.

**Review by zherman0 — APPROVED** [review]

**Review by spadgett — APPROVED** [review]
/lgtm

File: frontend/packages/console-dynamic-plugin-sdk/scripts/generate-doc.ts
Comment (andrewballantyne) [inline frontend/packages/console-dynamic-plugin-sdk/scripts/generate-doc.ts:96]:
```suggestion
  const link = `../../../${urlPath}`;
```
Would this not always work? The docs are generated on a specific path (`frontend/packages/console-dynamic-plugin-sdk/docs/api.md`) -- couldn't we just work our way back to just before `packages|public` -- then let your `urlPath` build us up from there?

File: README.md
Comment (TheRealJon) [reply README.md:474]: They are fully qualified because they are meant to link to the source code and not the README. Some packages don't have READMEs, but I still wanted to provide a link to something. Also, github pages is deploying from our master branch and deploys changes immediately, so we shouldn't have any issue with broken links as long as we update the README when a package leaves the repo.

File: frontend/packages/console-dynamic-plugin-sdk/scripts/generate-doc.ts
Comment (TheRealJon) [reply frontend/packages/console-dynamic-plugin-sdk/scripts/generate-doc.ts:96]: This would only work if we were linking to other markdown files. I don't think GitHub Pages will let you link to any old file in the repo.

File: README.md
Comment (andrewballantyne) [reply README.md:474]: We most definitely will have broken links with this model 🤔 

Release 4.11 => New Pages with "master" link
Release 4.12 => New Package comes along, new "master" link
Release 4.13 => pipelines-plugin goes Dynamic and deletes their references in OpenShift Console; Breaks 4.11 and 4.12 release branches, their link will go to a 404 page.

Relative links do not need to go to a file, they can go to a folder too. An example (click the top link): https://github.com/andrewballantyne/code-examples/blob/test-github-path/README.md

File: frontend/packages/console-dynamic-plugin-sdk/scripts/generate-doc.ts
Comment (andrewballantyne) [reply frontend/packages/console-dynamic-plugin-sdk/scripts/generate-doc.ts:96]: "old file"? Everything is relative to the branch you're on. So if I am looking at a partial commit, all links that do not override will navigate me around in that commit. I'm fairly certain this is how it works. Have you seen actual issues or is this an assumption? Git is just a series of diffs on top of each other... the branch is just a pointer in the tree. I can't think of a reason this won't work 🤔 I have no experience with github pages, but I assume it works the same way as a repo.

File: README.md
Comment (andrewballantyne) [reply README.md:474]: Ah -- right you are. https://andrewballantyne.github.io/code-examples does not work without a readme at the other end. 🤔

Comment (andrewballantyne) [reply README.md:474]: I stubbed an empty readme and it worked again. Okay... my recommendation is not to highlight packages that do not want to describe themselves.

I see nothing but a nightmare of maintenance by adding branch names.

File: frontend/packages/console-dynamic-plugin-sdk/scripts/generate-doc.ts
Comment (andrewballantyne) [reply frontend/packages/console-dynamic-plugin-sdk/scripts/generate-doc.ts:96]: Yup this won't work on GitHub pages. You cannot link to a file 😒 Going to `master` will cause problems, almost guaranteed. Files move, get removed, change in a fashion that is not going to live well in documentation. Navigating to the latest is almost guaranteed to fail after a short period of time. 

I'm unsure if GitHub pages will allow us to link to the code 🤔 Maybe we should just start reporting errors to all things that do not have comments for them. Omit them from the readout and present errors that we have to fix before it publishes to our official docs. 

Can we instead cause a build failure somehow 🤔 If we crash the generate-doc when there is no documentation for something it may get horribly inconvenient to development cycles. Maybe write the failures out to a new file & write a check for it. We should be able to verify this on the CI as it builds and generates everything we have.

Comment (TheRealJon) [reply frontend/packages/console-dynamic-plugin-sdk/scripts/generate-doc.ts:96]: I think this was initially implemented as a stop-gap until we could fill out all of the exposed APIs with TSDoc. 

That being said, I think it's okay that we fallback to a link to the API's source file if we don't have any TSDoc. IMO, that's better than omitting it from the docs entirely. The links should never be out of sync because we are using typescript compiler or require.resolve to dynamically build these paths.

Comment (TheRealJon) [reply frontend/packages/console-dynamic-plugin-sdk/scripts/generate-doc.ts:96]: > "old file"?

Sorry, that was just an idiom 😆 "any-old-file"

File: README.md
Comment (TheRealJon) [reply README.md:474]: I see what you mean. Hmm, okay, maybe we only provide links to packages with readme's then.

File: frontend/packages/console-dynamic-plugin-sdk/scripts/generate-doc.ts
Comment (TheRealJon) [reply frontend/packages/console-dynamic-plugin-sdk/scripts/generate-doc.ts:96]: > ~That being said, I think it's okay that we fallback to a link to the API's source file if we don't have any TSDoc. IMO, that's better than omitting it from the docs entirely. The links should never be out of sync because we are using typescript compiler or require.resolve to dynamically build these paths.~

Hmm, okay, I think I'm getting what you mean now. Scratch this. I'm thinking in terms of what we are showing in GitHub Pages, but if someone is reading the markdown independently in a release branch, the links would be outdated. Maybe we just use a relative path without a hyperlink? Just say 'see the implementation at `frontend/packages/src/foo/bar.ts'.

**Review by andrewballantyne — APPROVED** [review]
My original issues were around how we could end up with multiple doc versions. a 4.11 pointing at `master` would be a bad thing. But it's just the latest docs.

This is a bit of an oddity... GitHub pages is a single version -- the latest. So any link to `master`, as long as we keep it up to date, will work.

However anyone looking at the READMEs in previous version, they will get an undesired link reference.

I think we should go with this, and start thinking on how we can get the docs to be more future proof.

**Comment by TheRealJon** [comment]
QE Approver
/assign @yapei

Docs Approver:
/assign @opayne1

PX Approver:
/assign @RickJWagner

@spadgett @andrewballantyne I've updated all of the github links to the release-4.11 branch. This will need another lgtm.

File: README.md
Comment (yapei) [inline README.md:475]: a dot `.` is needed so that it can link to correct page

**Review by andrewballantyne — APPROVED** [review]
/lgtm

**Comment by TheRealJon** [comment]
@yapei I've updated per your comment

**Review by yapei — APPROVED** [review]
/label qe-approved

**Pull request merged**
